operator/pkg/ciliumendpointslice/reconciler.go

```diff
@@ -68,14 +73,43 @@ func (r *reconciler) reconcileCESUpdate(cesToUpdate string) (err error) {
 		return
 	}
 
-	// Call the client API, to Create CESs
-	if retCES, err = r.client.CiliumEndpointSlices().Update(
-		context.TODO(), ces, metav1.UpdateOptions{}); err != nil {
+	wasConflict := false
+
+	// Call the client API, to Update CESs'
+	if err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		retCES, err = r.client.CiliumEndpointSlices().Update(
+			context.TODO(), ces, metav1.UpdateOptions{})
+		if err != nil {
+			log.WithError(err).WithFields(logrus.Fields{
+				logfields.CESName: cesToUpdate,
+			}).Info("Unable to update CiliumEndpointSlice in k8s-apiserver, will retry if it's conflict")
+			if errors.IsConflict(err) {
+				wasConflict = true
+				// Update metadata of the object from store on conflict
+				obj, exists, getErr := r.cesStore.GetByKey(cesToUpdate)
+				if getErr != nil {
+					return getErr
+				}
+				if !exists {
+					return err
+				}
+				ces.ObjectMeta = obj.(*cilium_v2.CiliumEndpointSlice).ObjectMeta
```

Comment:
This looks a little questionable. Might want to see if you can update the obj.Spec/obj.Status instead, otherwise you might overwrite values added by another update.

Reply:
I think it's fine because the operator doesn't want to modify CES ObjectMeta.
The operator creates CESs and then only modifies the Endpoints field.

The only remaining concern is that updates will still fail if another source (other than operator) modifies the Endpoints field, but the aim of this PR is to fix the issue of operator getting stuck on its own, just as a part of the k8s system.

Reply:
I'm not worried about the ObjectMeta data, I'm worried that it's overwriting the entire rest of the object without concern for it's previous values.

Reply:
I don't understand it. I intended to overwrite ObjectMeta only and I thought this is what I did.
This is the same as updateCESInCache code called later:
```
func (c *cesMgr) updateCESInCache(srcCES *cilium_v2.CiliumEndpointSlice, isDeepCopy bool) {
    if ces, ok := c.desiredCESs.getCESTracker(srcCES.GetName()); ok {
        ces.backendMutex.Lock()
        defer ces.backendMutex.Unlock()
        if !isDeepCopy {
            ces.ces.ObjectMeta = srcCES.ObjectMeta
        } else {
            ces.ces = srcCES
                for _, cep := range ces.ces.Endpoints {
                    // Update the desiredCESs, to reflect all CEPs are packed in a CES
                    c.desiredCESs.insertCEP(GetCEPNameFromCCEP(&cep, ces.ces.Namespace), srcCES.GetName())
                }
            }
        } else {
            log.WithFields(logrus.Fields{
            logfields.CESName: srcCES.GetName(),
            }).Debug("Attempted to updateCESInCache non-existent, skipping.")
    }
}
```

Reply:
Ah, you are worried that in the apisever we ignore all other possible changes.

What is the case in which we shouldn't do this?
Operator should be the only controller for CES.

If some client changes k8s.Endpoints etc the controllers owning them will revert such changes. The same would happen here (to some extent) with the CiliumEndpointSlice.

Long term I think the operator should always reconcile the full state of CiliumEndpointSlice and make sure they are exactly as operator expects them to be.

Reply:
Since the operator "owns" the CiliumEndpointSlice, it should just unset ResourceVersion from the Update request. Then the apiserver will never reject it for concurrency issues, and there will be no need to retry.

Reply:
> Operator should be the only controller for CES.

If we run multiple workers (go routines) (I don't know if this is the case), then the operator could still be racing itself.

The problem here is that we're modifying our cache, which makes the control flow of state updates bi-directional. In a controller, we would ideally not mutate the cache directly but wait for the reconciliation and update after the cache is updated.

Reply:
> In a controller, we would ideally not mutate the cache directly but wait for the reconciliation and update after the cache is updated.

totally agree, but my observation in the cilium codebase is that there is no implementation of the reconciler pattern (unfortunately) and everything is based on an event based pattern using `Upsert`calls and caches on top of caches (like this one, if we already have the informer cache with the state of the world we should not need an additional one)

Is the existing code implementing reconcile loops?

Reply:
I'd like to go into full reconciler direction:
On event and every some time operator could compare expected state it knows based on CEPs, IDs with the existing state from the informer cache and make changes.

We are not there yet. Currently if CES is updated in apiserver but not in the cache (updated by other client or updated by operator but not updated cache (e.g. server updated but returned error so operator didn't updated cache) then the state for such single CES will be broken and won't reconcile until operator is restarted. It won't be able to remove old endpoints from the slice or add new endpoints to the slice.

Reply:
> Since the operator "owns" the CiliumEndpointSlice, it should just unset ResourceVersion from the Update request.

resource version MUST be set on an Update 😄 , you can trigger a CreateOnUpdate for some specific objtect but that is not for update the object, @squeed are you referring to some special case?

The canonical solution to solve multiple controllers ownership in kubernetes is to use server side apply https://kubernetes.io/docs/reference/using-api/server-side-apply/ but I don't whink we have this problem here, I think we should try with a PATCH instead of an UPDATE

Reply:
> resource version MUST be set on an Update 😄 

It doesn't have to be :-). It's just good practice in the event of concurrency.

Given that we don't have multiple controllers, unsetting ResourceVersion is the functional equivalent to setting Force in a server-side apply. In this case, it is the safe thing to do.

Reply:
> keep the metadata from the apiserver.

The apiserver will do that for you whether or not the ResourceVersion is set. Either the fields are not writable by clients and ignored (e.g. creationDate, UID, etc), or they are writable and need to be provided on Update.

Reply:
seem Casey is right, but follow the discussion with api-machinery folks https://kubernetes.slack.com/archives/C0EG7JC6T/p1687868578131159 so we fully understand this

Reply:
Ok, Casey and I asked for help and we found the explanation.

It turns out some types allow "UnconditionalUpdates" that is what Casey describe in his example, unfortunatly this is not available for CustomResources, and the recommendation is to use Patches

https://github.com/kubernetes/kubernetes/issues/83222#issuecomment-553408478

Reply:
Yup, I'm wrong, but for an interesting reason :-).

Reply:
I was right by chance, I didn't knew about "UnconditionalUpdates" , lucky we have Stefan to clarify these things :)

Reply:
How can we resolve this conversation.

I see two options:
1) The way I did it. Keep current behavior until conflict, update metadata on conflict.
It's not nice and clean (I hope I'll have time to implement proper reconciliation logic but it would be much bigger change). The biggest advantage is that nothing is changed until conflict (which is really rare).

2) Use PATCH instead of UPDATE. It will look better probably. There will be no conflicts. But it would require much bigger change to the way how we update CESs in all the cases.

@aojea @squeed WDYT?

Reply:
hmm ... there is a reconcile loop that will retry later

Or I miss something or for me the bug is that the cache is updated after doing the Update operation, and checking the `updateCESInCache` logic I think that this has to happen before, since is the one that gets the desired state

```diff
diff --git a/operator/pkg/ciliumendpointslice/endpointslice.go b/operator/pkg/ciliumendpointslice/endpointslice.go
index ce68044797..c515c9844e 100644
--- a/operator/pkg/ciliumendpointslice/endpointslice.go
+++ b/operator/pkg/ciliumendpointslice/endpointslice.go
@@ -351,6 +351,8 @@ func (c *CiliumEndpointSliceController) syncCES(key string) error {
                                return err
                        }
                } else {
+                       // Update local datastore CES with latest object metadata values.
+                       c.Manager.updateCESInCache(ces, false)
                        if err := c.reconciler.reconcileCESUpdate(key); err != nil {
                                return err
                        }
```

Reply:
I don't think it would work better.

It would fix the problem with conflict eventually but it would introduce more problem.
It would always use metadata from the watcher cache so the metadata would be outdated after update before watch gets the event.

This is happening right now with creates. We determine create vs update based on the watch cache, we create if CES is not present in the watch cache. So we observed it happening that we create the CES and operator wants to update it before the watcher obeserves the event so it tries to create again and fails.

Reply:
I could move my change to the handleErr - I think it makes more sense, will update the PR soon

Reply:
> It would always use metadata from the watcher cache so the metadata would be outdated after update before watch gets the event.


that is how kubernetes controllers work, there is no pass through cache and you have to operate with the state of the informer, it eventually be consistent

Reply:
> This is happening right now with creates. We determine create vs update based on the watch cache, we create if CES is not present in the watch cache

that is the part I don't understand, why are those two different operations?

Reply:
PTAL

I changed the place for handling conflict errors to errorhandling in endpointslices.